docs/bugs.json

```
@@ -1,4 +1,14 @@
 [
+    {
+        "uid": "SOL-2022-2",
+        "name": "NestedCallataArrayAbiReencodingSizeValidation",
+        "summary": "ABI-encoding values in calldata that contain nested dynamic arrays sometimes read beyond ``calldatasize()`` instead of reverting.",
```

```suggestion
        "summary": "ABI-encoding values in calldata that contain nested dynamic arrays sometimes reads beyond ``calldatasize()`` instead of reverting.",
```

---

Or how about this?
```suggestion
        "summary": "ABI-reencoding of nested dynamic calldata arrays may not detect invalid size and read beyond array bounds.",
```
It may happen at any point inside, crossing into another nested array. It's not just about reading past calldata, right?

---

``read`` was meant to be past tense :-).
``"ABI-reencoding of nested dynamic calldata arrays did not always perform proper size checks against the size of calldata and read beyond ``calldatasize``."``

It's actually not the array bounds that are violated - that'll fortunately never happen, it'd be much more serious. It's just that the array may not fit into the explicitly passed calldata entirely.

---

> It's actually not the array bounds that are violated - that'll fortunately never happen, it'd be much more serious. It's just that the array may not fit into the explicitly passed calldata entirely.

Right. I incorrectly remembered that we're checking if length fits in the available space but we're actually only checking if it does not overflow past the end of calldata. So the description does match the code.

But in that case - do we even validate that array items do not overlap other items or go past the space allocated for the whole array? I.e. what if the space allocated for an item as indicated by head offsets is too short compared to item lengths?

---

> `read` was meant to be past tense :-).

Fair enough. Looks like I just read it wrong :)

---

> But in that case - do we even validate that array items do not overlap other items or go past the space allocated for the whole array? I.e. what if the space allocated for an item as indicated by head offsets is too short compared to item lengths?

- do we even validate that array items do not overlap other items?

Not sure what you mean exactly, but generally encodings with overlapping arrays are valid, i.e.
```
contract C { function f(uint[] calldata x, uint[] calldata y) external { ... } }
// ----
// f(uint256[],uint256[]): 0x40, 0x40, 3, 1, 2, 3
```
is just fine and passes the very same array as ``x`` and ``y``.

- go past the space allocated for the whole array?

Also not sure what you mean, but there is no such thing as "space allocated for the array" - i.e. the following is also perfectly fine:
```
contract C { function f(uint[] calldata x, uint[] calldata y) external { ... } }
// ----
// f(uint256[],uint256[]): 0x80, 0x40, 3, 1, 2, 3, 4
```
which will pass ``[3, 4]`` as ``x`` and ``[1, 2, 3]`` as ``y``.
(Yes, ABI-encoding is weird ;-))

---

>  generally encodings with overlapping arrays are valid

Oh, that's interesting to know. I expected these checks to be much stricter because otherwise you can have a situations similarly "bad" to what we have with this bug, just without going past the calldata. I.e. if the client computes the length incorrectly and it's too long, the decoder will just grab the data belonging to another item. Just like here you can have wrong length and grab extra zeros from the empty space after calldata and here we consider it a bug - even though the result is actually more predictable and potentially even safer. I could actually see someone trying to be clever and using it intentionally to avoid paying for a bunch of trailing zeros :)

> Also not sure what you mean, but there is no such thing as "space allocated for the array"

Not sure how to call it properly but I was thinking about the space between the offset of your item and the offset of the next one. But you're right, the offsets do not even have to increase monotonically and very weird cases are possible.

It's just a bit odd that we care about not going beyond the end of calldata if we do not care about restricting these weird cases. Re-encoding them runs into the same potential problem with proxies you mentioned on the chat where data encoded again will be different from what you get as input.

---

> It's just a bit odd that we care about not going beyond the end of calldata if we do not care about restricting these weird cases. Re-encoding them runs into the same potential problem with proxies you mentioned on the chat where data encoded again will be different from what you get as input.